pkg/policy/selectorcache.go

```diff
@@ -0,0 +1,336 @@
+		key:              key,
+		users:            make(map[CachedSelectionUser]struct{}),
+		selector:         selector,
+		cachedSelections: make(map[identity.NumericIdentity]struct{}),
+	}
+
+	// Add the initial user
+	newIDSel.users[user] = struct{}{}
+
+	// Find all matching identities from the identity cache.
+	for numericID, lbls := range sc.idCache {
+		if selector.Matches(lbls) {
+			newIDSel.cachedSelections[numericID] = struct{}{}
+		}
+	}
+	// Create the immutable slice representation of the selected
+	// numeric identities
+	newIDSel.updateSelections()
+
+	// Note: No notifications are sent for the existing
+	// identities. Caller must use GetSelections() to get the
+	// current selections after adding a selector. This way the
+	// behavior is the same between the two cases here (selector
+	// is already cached, or is a new one).
+
+	sc.selectors[key] = newIDSel
+	return newIDSel, true
+}
+
+// RemoveIdentitySelector removes CachedSelector for the user.
+func (sc *SelectorCache) RemoveIdentitySelector(user CachedSelectionUser, selector CachedSelector) {
+	key := selector.String()
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+	idSel, exists := sc.selectors[key]
+	if exists {
+		if idSel.removeUser(user) {
+			delete(sc.selectors, key)
+		}
+	}
+}
+
+// UpdateIdentities propagates identity updates to selectors
+func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+
+	// Update idCache so that newly added selectors get
+	// prepopulated with all matching numeric identities.
+	for numericID := range deleted {
+		if _, exists := sc.idCache[numericID]; exists {
+			delete(sc.idCache, numericID)
+		} else {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Deleting a non-existing identity")
+		}
+	}
+	for numericID, lbls := range added {
+		if _, exists := sc.idCache[numericID]; exists {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Adding an existing identity")
+		}
+		sc.idCache[numericID] = lbls
+	}
+
+	// Iterate through all locally used identity selectors and
+	// update the cached numeric identities as required.
+	for _, sel := range sc.selectors {
+		var adds, dels []identity.NumericIdentity
+		switch idSel := sel.(type) {
+		case *labelIdentitySelector:
+			for numericID := range deleted {
+				if _, exists := idSel.cachedSelections[numericID]; exists {
+					dels = append(dels, numericID)
+					delete(idSel.cachedSelections, numericID)
+				}
+			}
+			for numericID, lbls := range added {
+				if _, exists := idSel.cachedSelections[numericID]; !exists {
+					if idSel.selector.Matches(lbls) {
+						adds = append(adds, numericID)
+						idSel.cachedSelections[numericID] = struct{}{}
+					}
+				}
+			}
+			if len(dels)+len(adds) > 0 {
+				idSel.updateSelections()
+				idSel.notifyUsers(adds, dels)
+			}
+		case *fqdnSelector:
+			// TODO
+		}
```

Comment:
@jrajahalme thoughts on this? ^ we may not even need this switch statement in the FQDN case since there is no logical way of corresponding a CIDR identity with a given FQDN by virtue of the fields of the identity itself at this current time - this relation has to come from the DNS proxy.